File: problems/This_is_coding_test/2_X/12_9 문자열 압축.py
# Counter도 안 됨. 중간에 낄 수 있음.
# 처음부터 끝까지 탐색 필요.


def solution(s):
    cur_total_count = 1000

    # XXX: 최대 범위는 절반까지
    mid = len(s) // 2
    for length in range(1, mid + 1):
        total_count = 0
        start = 0
        count = 1

        while start < len(s):
            # XXX: Python은 범위에 관대하니, 다른 걸로 풀기
            prev = s[start : start + length]
            nxt = s[start + length : start + length * 2]

            if prev == nxt:
                count += 1
            else:
                # XXX: 뒤에 오는 if문의 범위는 ()로 묶어야 함.
                total_count += len(prev) + (len(str(count)) if count > 1 else 0)
                count = 1
                prev = nxt

            start += length

        cur_total_count = min(cur_total_count, total_count)

    answer = cur_total_count
    return answer


print(solution("a"))

File: problems/This_is_coding_test/2_X/12_9 문자열 압축.js
/**
 * @param {string} s
 * @param {number} range
 * @returns string
 */
function compress(s, range) {
  let totalCount = 0;
  let count = 1;

  let start = 0;
  while (start < s.length) {
    // XXX: JavaScript에서도 substring을 쓰는데, Python처럼 관용적임...
    let prev = s.substring(start, start + range);
    let next = s.substring(start + range, start + range * 2);

    if (prev == next) {
      count += 1;
    } else {
      totalCount += prev.length + (count > 1 ? String(count).length : 0);
      count = 1;
      prev = next;
    }

    start += range;
  }

  return totalCount;
}

/**
 * @param {string[]} inputs
 */
function solution(inputs) {
  const s = inputs[0].toString();
  let minTotalCount = s.length;

  const mid = Math.floor(s.length / 2);
  for (let range = 1; range <= mid; range++) {
    const compressedCount = compress(s, range);
    minTotalCount =
      minTotalCount < compressedCount ? minTotalCount : compressedCount;
  }

  return minTotalCount;
}

const rl = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

let inputs = [];
rl.on("line", (line) => {
  inputs.push(line);
}).on("close", () => {
  const answer = solution(inputs);
  console.log(answer);
  process.exit();
});

/**
 * abcabcabcabcdededededede
 * # 14
 */
